pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1340,18 +1341,42 @@ private void asyncDeleteCursorWithCleanCompactionLedger(PersistentSubscription s
             return;
         }
 
-        currentCompaction.handle((__, e) -> {
-            if (e != null) {
-                log.warn("[{}][{}] Last compaction task failed", topic, subscriptionName);
+        // Avoid concurrently execute compaction and unsubscribing.
+        synchronized (this) {
+            if (!disablingCompaction.compareAndSet(false, true)) {
+                unsubscribeFuture.completeExceptionally(
```

Comment:
this should be protected with `if (!unsubscribeFuture.isDone()) {` so that creating the exception could be avoided.

Reply:
Could you mention the code line? I could not find it.